[Cards and Buttons] "Vertical lift" effect on hover

kjellr opened the pull request:
### PR Summary: 

Adds a simple "raise" effect to buttons and cards on hover. 

### Why are these changes introduced?

Closes https://github.com/Shopify/dawn/issues/2574. 

### What approach did you take?

Piggybacks on the work done in https://github.com/Shopify/dawn/pull/2452. Uses the same control in Theme Settings, and applies a different set of rules to just simply raise the elements on hover. 

### Visual impact on existing themes

When merchants visit Theme Settings > Animation, and choose the "Vertical lift" hover effect, cards and buttons will each. get a little "bump" effect on hover: 

https://user-images.githubusercontent.com/1202812/234607022-9aacddad-b9bd-4355-94db-4d3dc7d8d8ec.mp4

### Testing steps/scenarios

- Visit the Theme Settings > Animations
- Select "Vertical lift" from the "Hover Effect" dropdown. 
- Note that the raise effect is applied to buttons and cards throughout the storefront. 

### Demo links

- [Store](https://os2-demo.myshopify.com/?preview_theme_id=139846287382)
- [Editor](https://os2-demo.myshopify.com/admin/themes/139846287382/editor)

### Checklist
- [x] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [x] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [x] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [x] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [x] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [x] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [x] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

melissaperreault commented on assets/base.css:
```diff
@@ -3418,4 +3418,30 @@ details-disclosure > details {
+  .animate--hover-raise .customer button,
+  .animate--hover-raise .shopify-payment-button__button,
+  .animate--hover-raise .deferred-media__poster-button {
+    transition: transform var(--duration-long) cubic-bezier(0,0,.35,1);
```
I wonder if for this one we could pick the `--duration-default` instead of the long one. That could make this option snappier!

kjellr replied:
I adjusted this in https://github.com/Shopify/dawn/pull/2575/commits/d3718a033d1091751519564f342afabe69fec0dc. It felt too quick for larger elements like cards, but it felt okay for buttons. So I've only applied the `--duration-default` to buttons for now. I'm not sure how I feel about it yet — let me know what you think.

stufreen commented on assets/base.css:
```diff
@@ -3418,4 +3418,30 @@ details-disclosure > details {
+  .animate--hover-raise .customer button,
+  .animate--hover-raise .shopify-payment-button__button,
+  .animate--hover-raise .deferred-media__poster-button {
+    transition: transform var(--duration-long) cubic-bezier(0,0,.35,1);
```
If we're using the easing in multiple places we should put it in a variable, e.g. `--ease-out-slow`

kjellr replied:
Done in https://github.com/Shopify/dawn/pull/2575/commits/ca9883a554ab9b087cadfbd8b19ffc181b70505f.

stufreen reviewed (approved):
Code LGTM with one suggestions

kimberlyoleiro commented:
I agree with @melissaperreault here. The grow + raise almost appears wavy to me. Like my eye is catching an odd movement. 

Notice it a lot on the product cards and the add-to-cart button on PDP

melissaperreault reviewed (commented):
Thanks Kjell! Adding notes, some we already discussed together:

- [ ] 3D lift left over: The video and 3D model buttons look like they will fall down.

https://user-images.githubusercontent.com/26230969/234673761-2b5087f9-e0c6-4d39-a066-e4128fa7908b.mp4

- [ ] Raise: The video and 3D model buttons only grow more instead of lifting up


https://user-images.githubusercontent.com/26230969/234674482-4b6af473-6400-42a5-9adb-d019bf2adf23.mp4

- [ ] Collage video block: no animations happens on the Play button BUT this preexisted this.
- [ ] Collage block's height has changed. Unsure when this got introduced but it's not on main when I checked. When you hide or remove the image block, the other blocks are shorter than expected. ([Video reference](https://screenshot.click/26-55-nil4e-nyd73.mp4))

kjellr commented:
@melissaperreault + @kimberlyoleiro I've pushed an update that removes the expanding border for buttons. Let me know if that seems a little better for you. I'm not totally sold on it yet — the hover effect seems pretty subtle without the border size change. 

https://user-images.githubusercontent.com/1202812/234677796-a39d4e0e-7b3d-4d34-a060-0214cc91fd3d.mp4

kjellr commented on locales/en.default.schema.json:
```diff
@@ -268,6 +268,9 @@
             "label": "None"
           },
           "options__2": {
+            "label": "Raise"
```
I changed the copy to "Vertical lift". I think it does work nicely alongside "3d lift". 👍

kjellr commented:
I fixed up the bugs that @melissaperreault found above, so this should be ready for another round of reviews! 

I changed the name of the feature to "Vertical lift", which I think pairs nicely with our existing "3d lift" nomenclature.

kjellr commented:
Thanks for finding those, @melissaperreault! All should be fixed now. 

I also sped up the card animation slightly. I'm very hesitant to go much faster than that since it can be quite jarring to move around a large element like a card on hover. I'm trying to also sync up with our scroll animations, which are a bit smoother and less quick than some of the pre-existing ones. 

This PR should be ready for another round of testing!

ludoboludo reviewed (commented):
Looking good 👍

kjellr commented on assets/base.css:
```diff
@@ -3315,4 +3317,36 @@ details-disclosure > details {
+  .animate--hover-vertical-lift .customer button:not([disabled]):hover,
+  .animate--hover-vertical-lift .shopify-payment-button__button:hover{
+    transform: translateY(-0.25rem);
+  }
```
Great suggestion! Added in 6568052, along with a light one for cards as well.

ludoboludo reviewed (approved):
It's lot of selectors repetition and would be nice to have something a bit cleaner but not sure how. There are specific properties for each scenario, etc. 

We should keep an eye on this to see how it behaves with our potential gradient fix we're looking into merging for the release branch.

melissaperreault commented:
Hey! Catching up on reviewing this piece! 😓 
have we removed the grow on hover when we lift or decided to keep it? Asking because it's perceivable and I can't track what was the latest update on that. Thanks!

melissaperreault reviewed (approved):
Took another look! thank you for all the care again! LGTM! 🎉

KaichenWang commented on assets/base.css:
```diff
@@ -3322,4 +3324,51 @@ details-disclosure > details {
+  .animate--hover-vertical-lift .button:not([disabled]):hover:after,
+  .animate--hover-vertical-lift .customer button:not([disabled]):hover:after,
+  .animate--hover-vertical-lift .shopify-payment-button__button:not([disabled]):hover:after {
+    --border-offset: 0.3px; /* Prevent the border from growing on buttons when this effect is on. */
```
Could we add to the comment and explain how we calculated the `0.3px` value?

KaichenWang reviewed (approved):
Looks great!!

kmeleta commented:
> I was about to merge this, but something weird is happening with outline buttons

Seems to be a mercifully harmless issue. Just a little custom properties gotcha (fixed here https://github.com/Shopify/dawn/pull/2575/commits/2522854a96932dbb0b3abc1f2a4db90d08d4434c). I don't have all the context around the decisions made here though, so please give it a look and verify outline buttons are hovering how we expect now @kjell
